### ISSUE TITLE
Weather API issues

### ISSUE THREAD

**Issue by AshishkrGoyal** [freeCodeCamp/freeCodeCamp#15878]
This is a follow up on https://github.com/freeCodeCamp/freeCodeCamp/issues/15871#issuecomment-328652410
There are two major issues with the weather API:

- [ ] Temperature issue. Many times, it shows minimum and maximum temp equal and it changes when you reload, which is not possible practically.
- [ ] Location is wrong. The location provided by the API is not perfect. In many cases, it's completely wrong for many instances.

Thanks :+1:

**Comment by rmonastra** [comment]
Post your code so we can see if we can figure out what's causing this problem.

**Comment by erictleung** [comment]
@AshishkrGoyal thanks for the issue. It is a bit unclear to me which freeCodeCamp API you're referring to. Are you talking about a challenge itself or the freeCodeCamp website? Also next time, please fill out the issue template so it helps us understand your issue more clearly. Thanks!

**Comment by AshishkrGoyal** [comment]
@erictleung  i am talking about weather API of freecodecamp.
Thanks 👍

**Comment by systimotic** [comment]
Thanks for reporting this @AshishkrGoyal! 

Useful links:
[Show the local weather challenge](https://www.freecodecamp.org/challenges/show-the-local-weather)
[API](https://fcc-weather-api.glitch.me/)
API source - can't find this yet (@MiloATH or anybody else, do you have a link for this?)

I can confirm that the API can act a bit weird. I'll share some links and tell you what's wrong.
- [?lat=30.2498&lon=30.3086](https://fcc-weather-api.glitch.me/api/current?lat=30.2498&lon=30.3086) Coordinates are rounded down to two decimals. This can give results in a location somewhat close to where you are, but not exactly where you are. (I got this example by coming up with random coordinates. It's close to correct)
- When I test an application using the API myself, the geolocation API gets my location correct within a distance of less than 10 meters. The rounded down results from the weather API were a little over 300 meters of (that's also about 300 yards). That's enough to show the wrong name in this case, but not enough to give the wrong weather results. The temperature displayed for my location is still correct. Depending on what the decimals of your location are, the distance could be a bit different, but not a lot. My latitude got almost no rounding, my latitude got near the maximum change from rounding.
- [?lat=9999&lon=9999](https://fcc-weather-api.glitch.me/api/current?lat=9999&lon=9999) It seems to default to the example coordinates when given invalid coordinates. This makes it difficult to see when you're doing something wrong. When the issue is actually caused by an invalid request, it can easily be seen as an invalid response. _Interestingly, it doesn't link to an icon when using invalid coordinates, and the weather information is different from when you actually use the default coordinates (compare the previous link to [?lat=35&lon=139](https://fcc-weather-api.glitch.me/api/current?lat=35&lon=139))._
- I think I have seen the issue where min and max temps were equal, but I haven't been able to reproduce it consistently yet. If anybody else can, please let me know! Looking at the code might help with this.

@AshishkrGoyal Can you share some of the examples that gave the wrong location for you?

**Comment by AshishkrGoyal** [comment]
@systimotic 
sure i will share with you my pen ASAP!!! 
thanks for concerning 👍

**Comment by AshishkrGoyal** [comment]
@systimotic it is link for my codepen here you can see the complete code for my project which is developed with using freecodecamp API.
https://codepen.io/AshishkrGoyal/full/GvBprw

Thanks 👍

**Comment by vkWeb** [comment]
@systimotic @AshishkrGoyal We have been receiving many complaints regarding the Weather API. I would like to talk about this in detail. 

![Weather Issue FCC - Desktop](https://i.imgur.com/lHTTfkf.png)
*Weather on Desktop - By FCC Weather API*


![Weather Issue FCC - Desktop](https://i.imgur.com/v2OWNoh.png)
*Weather on Desktop - By https://weather.com*


![Weather Issue FCC - Mobile Device](https://i.imgur.com/pbI3o0nm.png)
*Weather on mobile device - By FCC Weather API*


![Weather Issue FCC - Mobile Device](https://i.imgur.com/ojcoix2m.png)
*Weather on mobile device - By https://weather.com*


In all the cases above. Location shown by FCC weather API is accurate as compared with https://weather.com   
There is a slight variation in temperature which needs to be resolved.

**Conclusion**
- [ ] Resolve inaccurate temperature issue with FCC weather API. @systimotic @QuincyLarson 
- [x] Location shown by the FCC weather API is correct. @AshishkrGoyal 

@AshishkrGoyal Please try checking the weather on a GPS enabled device (eg. mobile device). You'll get accurate location details.
Please talk about that min. max. temperature issue in detail and if possible provide some relevant screenshots.

**Comment by MiloATH** [comment]
I made a repo for the glitch app. Please see https://github.com/MiloATH/WeatherAPI. I am more than willing to transfer ownership to freeCodeCamp if they want the project. I still need to add continuous integration and there are other things which could make this better. This is still a work in progress, but it is in the right direction. Also, it might be cool if we list campers projects on the landing page and let campers add their own (just a thought).

**Comment by QuincyLarson** [comment]
@MiloATH Awesome! Yes - once it's ready, load it into Glitch and share a link to it. We can then "remix" it (fork it) from your account to freeCodeCamp's Glitch account, then update the URL in the challenge. Please let me know once you're ready for us to do that. I would keep this as simple and functional as possible.

**Comment by MiloATH** [comment]
@QuincyLarson The biggest challenge with using "remix" is the API keys. What do you suggest? Won't each remix require it's own API keys?

**Comment by lindakatcodes** [comment]
So from what I can tell, the FCC weather API uses data from openweathermap. (Going off of Milo's repo he shared.) I think more of the problem with the temp & location/name inaccuracies simply has to do with openweathermaps' data. 

From my testing, the temp seems to be pretty accurate for places in the US, and off a few more degrees in other countries. They also round their coordinates down to two decimal places, which might explain some of the odd names that show up. I'm not sure there is much that can be done to fix these, as it has to do with openweathermaps' information, which we don't have any control over. (Did my testing using the api call in the browser, and comparing the info with weather.com's info.) 

As for the temp_min and temp_max sometimes showing the same info.....this has to do with the FCC API using the current weather call to openweathermap. Reading through their docs, the min/max in the current weather API is meant to show differences in different areas of big cities, and not as we might normally think of a daily min/max. It's an optional thing to show the developer some variances in that area. So, not reliable to use as a daily min/max.

**Comment by QuincyLarson** [comment]
@MiloATH To be more clear - campers wouldn't need to get an API key because they would hit an endpoint on our Glitch app, which would use its API keys to access Open Weather Map's API, then pass the results back to the camper.

So this would only be remixed once: from your original Glitch.com project over to freeCodeCamp's Glitch account (where it could be collectively maintained by contributors as needed). We'd add our own API keys. Then no further API keys would be necessary.

**Comment by barmetma** [comment]
@lindakatcodes just found the same info, looks like the name min and max temp is a bit confusing and the expected result would be the min and max temperature of the current day. But this is wrong.

https://openweathermap.org/forecast5#min

**Comment by QuincyLarson** [comment]
@lindakatcodes You may be able to adjust for some of these quirks on the Glitch.com app we're using as a sort of pass-through to the Open Weather Map API.

**Comment by lindakatcodes** [comment]
@QuincyLarson I'd be happy to help if I can! Just let me know what I can do and I'll look into it. I looked through the repo that @MiloATH posted above and didn't see anything that seemed like it would manipulate the open weather map info....but it's possible I missed something or didn't quite have the full picture.

**Comment by QuincyLarson** [comment]
@MiloATH Can you point to a good place in  https://github.com/MiloATH/WeatherAPI where @lindakatcodes might be able to experiment with manipulating the open weather map data before it gets passed to the end user?

**Comment by MiloATH** [comment]
@lindakatcodes Most of the data is requested and returned to the user using https://github.com/MiloATH/WeatherAPI/blob/master/server.js#L43. You could add a data manipulation function and call it right after line https://github.com/MiloATH/WeatherAPI/blob/master/server.js#L66.

**Comment by 0x0936** [comment]
Has consideration been made to using a different weather API, such as [Yahoo! Weather](https://developer.yahoo.com/weather/)? A jQuery plugin [simpleWeather.js](http://simpleweatherjs.com/) [cdn](https://cdnjs.com/libraries/jquery.simpleWeather) makes weather requests simple. Response includes high / low for the day, as well as a 10-day forecast.

Not included in simpleweather's response are the geographic coordinates (a decision made by the maintainers of simpleweather), but when using Yahoo's YQL console for weather.forecast, lat and lon are shown to have six digits of accuracy.

```
"item": {
     "lat": "37.371609",
     "long": "-122.038254",
```

```js
// example simpleweather usage
navigator.geolocation.getCurrentPosition(p => {
  const { latitude: lat, longitude: lon } = p.coords
  const location = `${lat},${lon}`
  $.simpleWeather({
    location,
    success: console.log
  })
})
```

**Comment by lindakatcodes** [comment]
I'm willing to look into this, see if it's more accurate than our current setup.

**Comment by lindakatcodes** [comment]
Alright, bit of a status update. Apologies if this is long!

I submitted a pull request to the repo @MiloATH made (https://github.com/MiloATH/WeatherAPI/pull/3) with a few small fixes. I added some extra information to the main API view that explains the min/max temp issue, and gives some extra links to open weather map's docs for weather descriptions and the current weather, if any other information is needed from that. Should hopefully clear some things up for folks.

Also included a small validation check in the main API JS file, to be certain a user's submitted lat/lon coordinates are valid options. If not, it returns an error telling them it's not in the valid range, and what that range should be. (Some folks in [Issue #15767 ](https://github.com/freeCodeCamp/freeCodeCamp/issues/15767) had some questions on this.)

I checked into the suggested simpleWeather/Yahoo idea above, and from my tests the names were closer to correct, but the temps were still sometimes off when compared to the weather channel's website.  Since I know, whenever the beta becomes the core, that this challenge is moving off the main map and into extra challenges, it didn't seem worth switching off of open weather map just for the names. Also, I'm not sure how the jQuery plugin would work, but Yahoo's weather API has a rate limit of 2,000 calls a day, whereas open weather map is only limited on a per-minute basis. Just seems more accessible for campers this way.

I'm honestly completely stumped on how to adjust the temp, and if it's even worth it. The initial issue that opened this thread was on the min/max idea, which I've added an explanation for on the API view. As far as it not matching what weather.com says is the current temp....I'm not sure there's anything feasible to do for that. I believe weather.com uses wunderground for their info, which has a rate limit that doesn't seem quite feasible for fCC. Just my opinion though, I might be wrong on that. Definitely interested in other ideas on this if anyone has some.

As for making the names better....I do think pulling a name based on the HTML5 geolocation info would be a better way to go. I think open weather map's places data just isn't quite as accurate, since it rounds the coordinates to 2 decimals. However, I couldn't decide on an API to use for that. There's quite a few options out there....yahoo, google maps, locationIQ, mapquest, and others....but they all have their own rate limits, and I just wasn't sure which option might best fit fCC. I'd be more than happy to implement an option if one were to be picked by @QuincyLarson and/or others that have a lot more say in this curriculum than I do.
